app/src/main/java/ch/epfl/sweng/rps/services/MultiplayerTicTacToe.kt

```diff
@@ -0,0 +1,23 @@
+package ch.epfl.sweng.rps.services
+
+import ch.epfl.sweng.rps.ui.tictactoe.TicTacToeFragment
+import kotlin.random.Random
+
+class MultiplayerTicTacToe(val view:TicTacToeFragment,player:MOVES) : TicTacToeGame(player){
+    override fun putChoice(square: Int) {
+        if(!gameRunning)
+            return
+        val cell = square % 3
+        val row = (square / 3)
+        if(matrix[row][cell] != MOVES.EMPTY)
+            return
+        matrix[row][cell] = player
+        view.updateUI(square, player)
+        calculate()
+        player = if(player == MOVES.CROSS) MOVES.CIRCLE else MOVES.CROSS
+    }
+    override fun gameOver(winner:MOVES){
+        gameRunning = false
+        view.gameOver(winner)
+    }
+}
```

Comment:
Are you sure this is the best approach ? I would prefer not change UI in the logic here, but instead only update the current state, and then ui listens for changes and updates the UI.

Reply:
I mean sure that would be an approach, tbf I don't think there is much difference just that implementing an observer in the ui is more painful than just providing a callback where ui has freedom to do what it pleases.

Reply:
Sure that would be an approach too! I choose this one because in your approach I needed to have an observer for the TTT state which is a bit more problematic to implement correctly while with a callback based approach it was super straightforward and easy to implement.